clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -479,6 +479,12 @@ private static int parseAcks(String acksString) {
             // handling exceptions and record the errors;
             // for API exceptions return them in the future,
             // for other exceptions throw directly
+        } catch (BufferExhaustedException e) {
+            this.errors.record();
+            this.metrics.sensor("buffer-exhausted-records").record();
+            if (this.interceptors != null)
+                this.interceptors.onSendError(record, tp, e);
+            throw e;
```

Comment:
I think this is gonna be a slight spec change. Before this, a `TimeoutException` thrown either by waiting metadata or by waiting buffer allocation were caught by the following clause for `ApiException`(since `TimeoutException` extends `RetriableException` which is an `ApiException`), so a `FutureFailure` returned instead of throwing, and the callback triggered too.

As the result of this change, two TimeoutException cases are treated differently:
- timeout occurred while waiting metadata update => callback called, `FutureFailure` returned instead of throw
- timeout occurred while waiting buffer allocation => callback not called, throw exception

This sounds confusing and inconsistent. I think we can either take 1. always return `FailureFuture` for `TimeoutException` or 2. always throw for timeout exception.
IMO, by design of KafkaProducer(method blocking is part of interface), the latter one makes more sense to me.
WDYT?